gencost_transp: gencost_errfile -> gencost_mask

timothyas opened the pull request:
## What changes does this PR introduce?
In pkg/ecco, transp cost functions now use "gencost_mask" rather than "gencost_errfile", which is consistent with boxmean cost functions. 

## What is the new behaviour 
Now users specify masks with gencost_mask in data.ecco. If errfile is specified, a message is sent to STDERR to tell the user to use gencost_mask (just as with boxmean). 

## Does this PR introduce a breaking change? 
No, it passed a quick test run.

edoddridge commented:
@gaelforget - are you able to review this?

gaelforget commented:
Sorry @timothyas for delay. Have not tested but seems fine. Some caveats:
- don't think this feature is actively tested anywhere beyond compiled and taf'ed in global_oce_cs32.
- adding warnings that trHeat and trSalt generally use inconsistent advection scheme (unless ENUM_CENTERED_2ND is actually used by model), don't include bolus velocity components, and don't include diffusion components either may be useful. 
- I would probably advise using m_horflux_vol in place of m_trVol since the former is actively tested in global_oce_cs32/input_ad.sens/ and treated more like m_boxmean is.

gaelforget commented:
@timothyas wrt "a re-write of gencost_transp would be ideal"

I believe that the m_horflux approach already is an appropriate rewrite. At the time I only ported the vol case because the heat and salt cases as implemented by @antnguyen13 seemed generally incorrect. But I think that adding heat, salt to the m_horflux approach would just be a matter of mapping the correct W/S intensive vector field to gencost_storefld in ecco_phys.F

timothyas commented:
Thanks for the clarification @gaelforget. At least this PR lets users know that horflux is the way to go (via error messages in STDERR). Perhaps some of these notes should be mentioned in the documentation as well. Let me know if you think any of this is necessary.

christophernhill commented:
@timothyas we were going to merge this, but it has a conflict with some changes related to mds => read_rec. We can't quite figure out how to resolve. Can you take a look and fix your branch - then we can merge.

timothyas commented:
It should all be fixed @christophernhill. My branch was not caught up with upstream/Master, but everything should be clean now. 

I also added the warnings I put in STDERR to the documentation, so people can see what they're getting into before running the model, I hope you guys agree. 

In fixing this conflict, I maybe went overboard and removed msktrvol[W/S], replaced them with gencost_mask. Now the reading is done in `ecco_readparms.F` rather than in `ecco_check.F` and in `ecco_phys.F` I moved the `trVol` etc. calculations to near `boxmean_horflux`. This was maybe overkill, but at first I was not sure where the READ_REC error was (I should have just rebased my branch before doing anything else). However, the behavior is exactly the same, and I would argue it is a bit cleaner. Again, I hope you agree but let me know if otherwise.

jm-c commented:
There is enough new/update here that it would be worth to have an other round of review (@gaelforget ).
Otherwise:
1) ecco_readparms is not the best place to load-in some 2d/3d mask (called before ini_model_io).
 the best place would be in "ecco_init_fixed" (but does not currently exist). Might be a good
opprotunity to create one.
2) did you try to trigger the warning message in eccho_check.F (lines 550-557) ? seems to me
that the printed line is getting a bit long. And I don't like strings that are left opened (line 551) and only closed by continuation line.

christophernhill commented:
@timothyas Jean-Michel had a question

timothyas commented:
Sorry for the delay! 
Re: @jm-c 
1. For the masks, I was following what was done for `m_boxmean` and `m_horflux`. If init_fixed is preferred I can implement that, with likely another iteration or so of comments/review.

2. I did trigger the warnings, and agreed they are a bit long. I could either omit them or make a smaller warning, mentioning the documentation I added which says the same thing. I think having these warnings somewhere is still a good idea, since for instance I had no idea that `gencost_transp` is not kept up with and I think other users may have the same confusion. Let me know what you prefer though. 

In either case, I will be sure to not leave strings open as in line 551 of `ecco_check.F` - thanks for the heads up.

jm-c commented:
@timothyas sorry for the delay.
Regarding the move to ecco_readparms.F, I was not aware that there was already some 2D/3D files loaded there. I still don't think it's a good place there, but could leave it like this (and fix this later).
And regarding error message, the content is OK all in one line will make it less readable; may be break it
in 4 lines, 1 at the top and then (1) (2) (3) ?

timothyas commented:
No worries at all @jm-c! Write statements are cleaned up.

jm-c commented:
LGM. Did not try to trigger errors (to check how err msg will come out) but re-run standard ADM testreport (+ global_oce_cs32 ) and was OK.

jm-c commented:
I think this is good for merging (may be "squash and merge"). @timothyas could you add 1 or 2 lines
of description in doc/tag-index ?

timothyas commented:
Hey @jm-c, I added a couple of lines but I wasn't sure how to do this - should these lines go above the "checkpoint 67c ..." line? 

Let me know, I can easily amend this commit. Cheers

jm-c commented:
@timothyas just moved up your comments in doc/tag-index. Waiting for travis test to finish and will
merge it.

timothyas commented:
@jm-c got it, noted for future reference. Thanks!
